rust/src/tree.rs

```diff
@@ -1,185 +1,1264 @@
+
+                    if shared_count == matched_node_text_count {
+                        // Full match with current node's text, continue to next node
+                        curr_idx += shared_count;
+                        prev = Arc::clone(&matched_node);
                     } else {
-                        curr_idx += prefix_len;
+                        // Partial match, stop here
+                        curr_idx += shared_count;
+                        prev = Arc::clone(&matched_node);
                         break;
                     }
                 }
-                None => {
+                Entry::Vacant(_) => {
+                    // No match found, stop here
                     break;
                 }
             }
         }
 
-        &input_ids[..curr_idx]
+        curr = prev.clone();
+
+        // Select the first tenant (key in the map)
+        let tenant = curr
+            .tenant_last_access_time
+            .iter()
+            .next()
+            .map(|kv| kv.key().to_owned())
+            .unwrap_or("empty".to_string());
+
+        // Traverse from the curr node to the root and update the timestamp
+
+        let timestamp_ms = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_millis();
+
+        if !tenant.eq("empty") {
+            let mut current_node = Some(curr);
+            while let Some(node) = current_node {
+                node.tenant_last_access_time
+                    .insert(tenant.clone(), timestamp_ms);
+                current_node = node.parent.read().unwrap().clone();
+            }
+        }
+
+        let ret_text = slice_by_chars(text, 0, curr_idx);
+        (ret_text, tenant)
     }
 
-    pub fn delete(&mut self, input_ids: &[u32]) {
-        let mut curr = &mut self.root;
-        curr.count -= 1;
+    fn leaf_of(node: &NodeRef) -> Vec<String> {
+        /*
+        Return the list of tenants if it's a leaf for the tenant
+         */
+        let mut candidates: HashMap<String, bool> = node
+            .tenant_last_access_time
+            .iter()
+            .map(|entry| (entry.key().clone(), true))
+            .collect();
 
-        let mut curr_idx = 0;
-        let input_ids_len = input_ids.len();
+        for child in node.children.iter() {
+            for tenant in child.value().tenant_last_access_time.iter() {
+                candidates.insert(tenant.key().clone(), false);
+            }
+        }
 
-        while curr_idx < input_ids_len {
-            let first_id = &input_ids[curr_idx];
+        candidates
+            .into_iter()
+            .filter(|(_, is_leaf)| *is_leaf)
+            .map(|(tenant, _)| tenant)
+            .collect()
+    }
 
-            if curr.children.contains_key(first_id) {
-                let child = curr.children.get(first_id).unwrap();
+    pub fn evict_tenant_data(&self, max_size: usize) {
```

Comment:
The priority queue (actually a linked list is better) can be maintained, rather than recompute each eviction time. The current implementation is also fine for fast move, since it is actually a lazy eviction and the complexity is amortized.

Reply:
could you explain how LL works?